### ISSUE TITLE
Default wire resistivity and readouts lead to unwieldy numbers

### ISSUE THREAD

**Issue by samreid** [phetsims/circuit-construction-kit-dc#125]
![image](https://user-images.githubusercontent.com/679486/28797085-4161f5d8-75fd-11e7-8163-6191e7123d67.png)

For ideal wires, this would read 12.0A but due to the resistivity it is reading 11.999A.  This seems liable to confuse middle schoolers and possible older students too.  We could address this by reducing the default resistivity or rounding some numbers in some displays.  Here's how long the wires have to be to reduce to 11.998

![image](https://user-images.githubusercontent.com/679486/28797153-86cc263e-75fd-11e7-823b-434810c03d15.png)

Discovered during https://github.com/phetsims/circuit-construction-kit-dc/issues/109

**Comment by arouinfar** [comment]
> For ideal wires, this would read 12.0A but due to the resistivity it is reading 11.999A. This seems liable to confuse middle schoolers and possible older students too.

@samreid I agree that this could be problematic for some students. I would recommend that you investigate lowering the default resistivity rather than rounding.

**Comment by samreid** [comment]
I took the default resistance from 1E-4 to 1E-10 and haven't seen any numerical issues yet.  And it handles default numbers nicely:

![image](https://user-images.githubusercontent.com/679486/28805345-a1b5790e-7626-11e7-8884-0a1c9f2f6b12.png)

It will need more testing.

**Comment by samreid** [comment]
I confirmed values are commensurate with the previous version, and tried to wreck the linear algebra, but couldn't find any problems after a few minutes of testing.

@arouinfar can you please take it for a test drive?
http://www.colorado.edu/physics/phet/dev/html/circuit-construction-kit-dc/1.0.0-dev.131/circuit-construction-kit-dc_en.html

**Comment by samreid** [comment]
Here's an extreme case I tested that didn't cause numerical issues:

![image](https://user-images.githubusercontent.com/679486/28805518-dc1ec018-7627-11e7-8961-76c106f63f5c.png)

**Comment by samreid** [comment]
Bumping priority since this is the only know issue before we start dev testing.

**Comment by arouinfar** [comment]
@samreid everything seems to be behaving. I also created a short with all 14 available batteries at max voltage, which seemed to run smooth.

The only side effect I'm noticing is that the shorts now have an extremely high current.
![image](https://user-images.githubusercontent.com/8419308/28835246-f81b92bc-76a1-11e7-825d-86d9429102d5.png)

Is the >10^10 behavior new? I know we did something similar for Black Box, but I hadn't seen it in DC before (probably because I never got the current that high).
![image](https://user-images.githubusercontent.com/8419308/28835284-1856a1b6-76a2-11e7-931a-8575361d70c8.png)

Seems like it would be good to discuss how to display such high currents in design meeting.

**Comment by arouinfar** [comment]
@samreid decreasing the resistivity be 6 orders of magnitude seems a bit drastic. Maybe we could go with 1E-7 or 1E-8 instead. I think it would be acceptable for circuits with long wires, such as the one in https://github.com/phetsims/circuit-construction-kit-dc/issues/125#issuecomment-319241884, to display 11.999 A, especially if would ease the instability in #126.

**Comment by samreid** [comment]
In #126 @arouinfar reported

This looks a bit off...

![image](https://user-images.githubusercontent.com/8419308/28835749-adfd3de6-76a3-11e7-97e6-a692a0dbbf4c.png)

**Comment by samreid** [comment]
It would be convenient if increasing the resistivity made all of the problems go away.  If we want something more guaranteed to address cases like the preceding comment, we could split apart the solution to disjoint "connected sets" and solve each independently.  That may take around 4 hours, but it would break the dependence between unconnected parts.

**Comment by samreid** [comment]
At resistivity of 1E-4, the decimal point drops at this wire length:
![image](https://user-images.githubusercontent.com/679486/28841025-2ead06da-76b6-11e7-9f6c-1894dbf788e0.png)

At 1E-5, the decimal point drops at this wire length:
![image](https://user-images.githubusercontent.com/679486/28841079-4f81901a-76b6-11e7-9eb9-789456e9d384.png)

At 1E-5, a high voltage battery with a default bulb reveals the resistivity right away:
![image](https://user-images.githubusercontent.com/679486/28841218-be25b708-76b6-11e7-8beb-6a1507595836.png)

**Comment by samreid** [comment]
I changed the default resistivity to 1E-5 and committed.

Testing again with 1E-10 and using a fake battery with 10000000V, I can get electrons to run through a disconnected wire:

![image](https://user-images.githubusercontent.com/679486/28841559-eaeacc1e-76b7-11e7-9206-24e69cd31b62.png)

**Comment by samreid** [comment]
@arouinfar here is a version with 1E-5 resistivity. http://www.colorado.edu/physics/phet/dev/html/circuit-construction-kit-dc/1.0.0-dev.133/circuit-construction-kit-dc_en.html

Can you test it out and see if it is ready for dev testing?  Or if we may need to something like https://github.com/phetsims/circuit-construction-kit-dc/issues/125#issuecomment-319458218 or anything else?

**Comment by arouinfar** [comment]
@samreid I haven't been able to reproduce https://github.com/phetsims/circuit-construction-kit-dc/issues/125#issuecomment-319432659 in dev.133, and the current looks good for standard components. I'd say you're good to go for dev testing, but I think we may want to discuss whether or not we're okay with rounding issues with the high V battery in design meeting.

**Comment by kathy-phet** [comment]
I note that in the Java version, the current decimal places only go to two decimals. This would solve all of the rounding issues for current, I think. 

Could we do two decimals?

Voltage drop shows 3 decimals, I think to capture the small voltage drops in the wires due to their resistance at high current.

**Comment by samreid** [comment]
Here's a snapshot from the latest version, showing 2 decimal points in the Voltmeter, and 3 in each ammeter:

![image](https://user-images.githubusercontent.com/679486/28853975-6883953c-76f1-11e7-97ee-6ac6a2a66be6.png)

I'm not sure if the voltmeter is supposed to be 3 decimal points, but it apparently has been 2 decimal places for quite a long time.  There is a note in the ammeter code that says the current shows 3 decimal places "so that current can still be seen with a glowing high-resistance bulb".  So I suspect the current meter used to have 2 decimal places.

@ariel-phet @arouinfar and @kathy-phet this is currently the only remaining issue before dev testing, please let me know your thoughts.

**Comment by kathy-phet** [comment]
Can we have a rule that shows only two decimals unless it is a current between 0.02 and 0.001 and the it starts showing 3 decimals?

Two decimals is easier to digest.

Perhaps we could have a similar rule for voltage?

I do recall we had made the voltmeter three decimals in the Java version to show the small voltage drop on the wires. I will have to play more with this and I'll try to get together with Amy to discuss.

Kathy

Sent from my iPhone

On Aug 1, 2017, at 7:45 PM, Sam Reid <notifications@github.com<mailto:notifications@github.com>> wrote:


Here's a snapshot from the latest version, showing 2 decimal points in the Voltmeter, and 3 in each ammeter:

[image]<https://user-images.githubusercontent.com/679486/28853975-6883953c-76f1-11e7-97ee-6ac6a2a66be6.png>

I'm not sure if the voltmeter is supposed to be 3 decimal points, but it apparently has been 2 decimal places for quite a long time. There is a note in the ammeter code that says the current shows 3 decimal places "so that current can still be seen with a glowing high-resistance bulb". So I suspect the current meter used to have 2 decimal places.

@ariel-phet<https://github.com/ariel-phet> @arouinfar<https://github.com/arouinfar> and @kathy-phet<https://github.com/kathy-phet> this is currently the only remaining issue before dev testing, please let me know your thoughts.

**Comment by samreid** [comment]
> Can we have a rule that shows only two decimals unless it is a current between 0.02 and 0.001 and the it starts showing 3 decimals?

That seems straightforward--less than an hour to implement for voltmeter and ammeters in the play area.  I'll wait for @kathy-phet and @arouinfar to confirm before I start.

**Comment by arouinfar** [comment]
I like @kathy-phet's suggestion. Go for it @samreid!

**Comment by samreid** [comment]
![image](https://user-images.githubusercontent.com/679486/28881543-b50ad5ae-7765-11e7-878d-2d60786b4f01.png)
![image](https://user-images.githubusercontent.com/679486/28881556-c15bd092-7765-11e7-97a1-8e0dbe70bed2.png)
![image](https://user-images.githubusercontent.com/679486/28881563-c874258c-7765-11e7-80aa-5b97231a6504.png)

**Comment by samreid** [comment]
Proposed implementation is here:
http://www.colorado.edu/physics/phet/dev/html/circuit-construction-kit-dc/1.0.0-dev.134/circuit-construction-kit-dc_en.html

@arouinfar can you please review and test?  And please point out anything else that should be fixed before dev testing.

**Comment by arouinfar** [comment]
@samreid I'm seeing some disagreement between the ammeters
![image](https://user-images.githubusercontent.com/8419308/28884896-f7bbd4e2-7770-11e7-972a-5d037f654814.png)

Looks like the non-contact ammeter's probe location makes a difference. The third digit appears when reading over a battery or resistor, but not a wire.
![image](https://user-images.githubusercontent.com/8419308/28885034-835eb85c-7771-11e7-83f0-7e8f3bdc61f2.png)
![image](https://user-images.githubusercontent.com/8419308/28885055-98d37376-7771-11e7-99ff-56c96ca26cb0.png)

**Comment by samreid** [comment]
@arouinfar how did you produce that bottom case?  I connected a normal battery and pencil like you did, then reduced the voltage to 4.0V and observed 0.013A throughout the circuit, even in the wires.

**Comment by samreid** [comment]
Likewise, for the top case I'm seeing agreement in the ammeters:
![image](https://user-images.githubusercontent.com/679486/28889983-e8360c06-7782-11e7-8def-c6e0141a58be.png)

**Comment by samreid** [comment]
@phet-steele determined that rotating the wire can cause the problem, and I was able to replicate the issue here.

**Comment by arouinfar** [comment]
Thanks @phet-steele. I've been having trouble consistently replicating.

**Comment by arouinfar** [comment]
@samreid on startup, I built the battery-pencil circuit and lowered the voltage. Non-contact ammeter sometimes displays the third digit, and sometimes doesn't.

**Comment by samreid** [comment]
Proposed fix is here:
http://www.colorado.edu/physics/phet/dev/html/circuit-construction-kit-dc/1.0.0-dev.135/circuit-construction-kit-dc_en.html

The wire direction was correlated with the problem, negative currents were not being given the correct number of decimal places, hopefully fixed now.  @arouinfar and @phet-steele can you please take a look?

**Comment by arouinfar** [comment]
I'm getting a consistent reading across all of these wires:
![image](https://user-images.githubusercontent.com/8419308/28892220-7624183a-778a-11e7-8110-aa3b25f6f1ba.png)

**Comment by arouinfar** [comment]
We discussed this issue during today's design meeting. A default resistivity of 10^-5 is reasonable, as is the thousandths place that appears in the meters when the magnitude is <.02.

The issue I raised in https://github.com/phetsims/circuit-construction-kit-dc/issues/125#issuecomment-319420027 is no longer relevant now that the resistivity has been increased.

@phet-steele can you verify this issue during the dev test?

**Comment by phet-steele** [comment]
I did verify this a bit ago...besides, the test in https://github.com/phetsims/circuit-construction-kit-dc/issues/125#issuecomment-319780842 is as good as any! 😄
